crates/server/src/server/server_future.rs

```diff
@@ -642,71 +672,141 @@ impl<T: RequestHandler> ServerFuture<T> {
+                    },
+                    _ = drain.clone() => {
+                        // A graceful shutdown was initiated. Break out of the loop.
+                        break;
+                    },
+                };
+
+                // verify that the src address is safe for responses
+                // TODO: we're relying the quinn library to actually validate responses before we get here, but this check is still worth doing
+                if let Err(e) = sanitize_src_address(src_addr) {
+                    warn!(
+                        "address can not be responded to {src_addr}: {e}",
+                        src_addr = src_addr,
+                        e = e
+                    );
+                    continue;
+                }
 
-                    inner_join_set.spawn(async move {
-                        debug!("starting quic stream request from: {src_addr}");
+                let handler = handler.clone();
+                let dns_hostname = dns_hostname.clone();
 
-                        // TODO: need to consider timeout of total connect...
-                        let result = quic_handler(handler, streams, src_addr, dns_hostname).await;
+                inner_join_set.spawn(async move {
+                    debug!("starting quic stream request from: {src_addr}");
 
-                        if let Err(e) = result {
-                            warn!("quic stream processing failed from {src_addr}: {e}")
-                        }
-                    });
+                    // TODO: need to consider timeout of total connect...
+                    let result =
+                        quic_handler(handler, streams, src_addr, dns_hostname, drain.clone()).await;
 
-                    reap_tasks(&mut inner_join_set);
-                }
+                    if let Err(e) = result {
+                        warn!("quic stream processing failed from {src_addr}: {e}")
+                    }
+                });
             }
+
+            reap_tasks(&mut inner_join_set);
+            Ok(())
         });
 
         Ok(())
     }
 
-    /// This will run until a background task of the trust_dns_server ends.
+    /// Triggers a graceful shutdown the server. All background tasks will stop accepting
+    /// new connections and the returned future will complete once all tasks have terminated.
+    pub fn shutdown(self) -> impl Future<Output = Result<(), ProtoError>> {
+        // Trigger the graceful shudown of all the server tasks.
+        self.drain.shutdown();
+
+        // Return the future that blocks until all tasks complete.
+        self.block_until_done()
+    }
+
+    /// This will run until all background tasks complete. If one or more tasks return an error,
+    /// one will be chosen as the returned error for this future.
     pub async fn block_until_done(mut self) -> Result<(), ProtoError> {
-        let result = self.join_set.join_next().await;
+        if self.join_set.is_empty() {
```

Comment:
Note to reviewers: I think this always should have blocked until all tasks are completed. With the addition of graceful shutdown, this needed to be corrected.

Reply:
Thanks for the heads up, this looks good.